test/js/symbol/binpack.test.js

```diff
@@ -4,97 +4,51 @@ var test = require('prova');
+        t.deepEqual(bp.allocate(10, 10), {x: 20, y: 0, w: 10, h: 10 }, 'third 10x10 bin');
         t.end();
     });
-    t.test('release single', function(t) {
-        var bp = new BinPack(10, 10);
-        var allocated = bp.allocate(10, 10);
-        var expected = {
-            x: 0,
-            y: 0,
-            w: 10,
-            h: 10
-        };
-        t.deepEqual(allocated, expected, 'packs a 10x10 square');
-        t.deepEqual(bp.allocate(10, 10), {
-            x: -1,
-            y: -1
-        }, 'cannot pack another');
-        var freed = bp.release(allocated);
-        t.deepEqual(freed, undefined, 'releases a 10x10 square');
+
+    t.test('packs larger bins on new shelf', function(t) {
+        var bp = new BinPack(64, 64);
+        t.deepEqual(bp.allocate(10, 10), {x: 0, y: 0, w: 10, h: 10 }, 'shelf 1, 10x10 bin');
+        t.deepEqual(bp.allocate(10, 15), {x: 0, y: 10, w: 10, h: 15 }, 'shelf 2, 10x15 bin');
+        t.deepEqual(bp.allocate(10, 20), {x: 0, y: 25, w: 10, h: 20 }, 'shelf 3, 10x20 bin');
         t.end();
     });
-    t.test('release many in same order', function(t) {
-        var bp = new BinPack(100, 100);
-        var allocated = [];
-        for (var i = 0; i < 20; i++) {
-            allocated.push(bp.allocate(10, 10));
-        }
-        for (i = 19; i >= 0; i--) {
-            bp.release(allocated[i]);
-        }
-        allocated = bp.allocate(10, 10);
-        var expected = {
-            x: 0,
-            y: 0,
-            w: 10,
-            h: 10
-        };
-        t.deepEqual(allocated, expected, 'packs a 10x10 square');
+
+    t.test('packs shorter bins on existing shelf, minimizing waste', function(t) {
+        var bp = new BinPack(64, 64);
+        t.deepEqual(bp.allocate(10, 10), {x: 0, y: 0, w: 10, h: 10 }, 'shelf 1, 10x10 bin');
+        t.deepEqual(bp.allocate(10, 15), {x: 0, y: 10, w: 10, h: 15 }, 'shelf 2, 10x15 bin');
+        t.deepEqual(bp.allocate(10, 20), {x: 0, y: 25, w: 10, h: 20 }, 'shelf 3, 10x20 bin');
+        t.deepEqual(bp.allocate(10, 9),  {x: 10, y: 0, w: 10, h: 9 }, 'shelf 1, 10x9 bin');
         t.end();
     });
-    t.test('release many in reverse order', function(t) {
-        var bp = new BinPack(100, 100);
-        var allocated = [];
-        for (var i = 0; i < 20; i++) {
-            allocated.push(bp.allocate(10, 10));
-        }
-        for (i = 0; i < 20; i++) {
-            bp.release(allocated[i]);
-        }
-        allocated = bp.allocate(10, 10);
-        var expected = {
-            x: 0,
-            y: 0,
-            w: 10,
-            h: 10
-        };
-        t.deepEqual(allocated, expected, 'packs a 10x10 square');
+
+    t.test('not enough room', function(t) {
+        var bp = new BinPack(10, 10);
+        t.deepEqual(bp.allocate(10, 10), {x: 0, y: 0, w: 10, h: 10 }, 'first 10x10 bin');
+        t.deepEqual(bp.allocate(10, 10), {x: -1, y: -1 }, 'not enough room');
```

Comment:
This can be a followup task, but `allocate` should return `null` or `undefined` rather than (-1, -1) to indicate failure.